File: TaskMaster/AppGlobals/AppToDoObjects.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using Newtonsoft.Json;
using Tags;
using TaskVisualization;
using ToDoModel;
using ToDoModel.Data_Model.People;
using UtilitiesCS;
using UtilitiesCS.EmailIntelligence;
using UtilitiesCS.Interfaces;
using UtilitiesCS.ReusableTypeClasses;
using UtilitiesCS.ReusableTypeClasses.Concurrent.Observable.Collection;
using UtilitiesCS.ReusableTypeClasses.Concurrent.Observable.Dictionary;
using UtilitiesCS.ReusableTypeClasses.Locking.Observable.LinkedList;
using UtilitiesCS.ReusableTypeClasses.SerializableNew.Concurrent.Observable;
using UtilitiesCS.Threading;

namespace TaskMaster
{
    public class AppToDoObjects(IApplicationGlobals parentInstance) : IToDoObjects
    {
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(
            System.Reflection.MethodBase.GetCurrentMethod().DeclaringType
        );

        public async Task LoadAsync(bool parallel = true)
        {
            if (parallel)
            {
                await LoadParallelAsync();
            }
            else
            {
                await LoadSequentialAsync();
            }
        }

        public async Task LoadParallelAsync()
        {
            var tasks = new List<Task>
            {
                LoadPrefixListAsync(),
                LoadDictRemapAsync(),
                LoadIdListAsync(),
                LoadProgramInfoAsync(),
                LoadProjInfoAsync(),
                LoadCategoryFiltersAsync(),
                LoadFilteredFolderScrapingAsync(),
                LoadFolderRemapAsync(),
                LoadFlagChangeTrainingQueueAsync(),
                LoadSelectFromListAsync(),
            };
            await Task.WhenAll(tasks);
        }

        public async Task LoadSequentialAsync()
        {
            await LoadPrefixListAsync();
            await LoadPeopleAsync();
            await LoadDictRemapAsync();
            await LoadIdListAsync();
            await LoadProgramInfoAsync();
            await LoadProjInfoAsync();
            await LoadCategoryFiltersAsync();
            await LoadFilteredFolderScrapingAsync();
            await LoadFolderRemapAsync();
            await LoadFlagChangeTrainingQueueAsync();
            await LoadSelectFromListAsync();
        }

        private readonly Properties.Settings _defaults = Properties.Settings.Default;

        private T Initialized<T>(T obj, Func<T> initializer)
        {
            obj ??= initializer.Invoke();
            return obj;
        }

        public IApplicationGlobals Parent { get; protected set; } = parentInstance;
        internal ISmartSerializableNonTyped SmartSerializable { get; set; } =
            new SmartSerializableNonTyped();
        internal Func<string, bool> FileExists { get; set; } = File.Exists;
        internal Func<string, string> ReadAllText { get; set; } = File.ReadAllText;

        private string _projInfo_Filename;
        public string ProjInfo_Filename =>
            Initialized(_projInfo_Filename, () => _projInfo_Filename = _defaults.FileName_ProjInfo);
        private ProjectData _projInfo;
        public IProjectData ProjInfo => Initialized(_projInfo, () => LoadProjInfo());

        private async Task LoadProjInfoAsync()
        {
            var outlookApplication = Parent.Ol.App;

            _projInfo = await Task.Run(() =>
            {
                if (Parent.FS.SpecialFolders.TryGetValue("AppData", out var appData))
                {
                    var proj = new ProjectData(
                        filename: _defaults.FileName_ProjInfo,
                        folderpath: appData
                    );
                    proj.Sort();
                    return proj;
                }
                else
                {
                    return null;
                }
            });

            if (_projInfo?.Count == 0 && outlookApplication is not null)
            {
                _projInfo.Rebuild(outlookApplication);
            }
        }

        private IProjectData LoadProjInfo()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("AppData", out var appData))
            {
                var projectInfo = new ProjectData(
                    filename: _defaults.FileName_ProjInfo,
                    folderpath: appData
                );
                if (projectInfo.Count == 0)
                {
                    projectInfo.Rebuild(Parent.Ol.App);
                }
                return projectInfo;
            }
            else
            {
                return null;
            }
        }

        private ScDictionary<string, string> _programInfo;
        public ScDictionary<string, string> ProgramInfo =>
            Initialized(_programInfo, LoadProgramInfo);

        private ScDictionary<string, string> LoadProgramInfo()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("AppData", out var appData))
            {
                return ScDictionary<string, string>.Static.Deserialize(
                    _defaults.FileName_ProgramDictionary,
                    appData
                );
            }
            else
            {
                return null;
            }
        }

        private async Task LoadProgramInfoAsync() => _programInfo = await Task.Run(LoadProgramInfo);

        //public ProgramData

        internal async Task LoadPeopleAsync() =>
            await Task.Run(
                async () =>
                {
                    if (Parent.IntelRes.Config.TryGetValue("People", out var config))
                    {
                        People = await SmartSerializable.DeserializeAsync(
                            config,
                            true,
                            () => new PeopleScoDictionaryNew(Parent)
                        );
                        People.Prefix = _prefixList?.Find(x =>
                            x.PrefixType == PrefixTypeEnum.People
                        );
                        People.CollectionChanged += People_CollectionChanged;
                    }
                    else
                    {
                        logger.Error("People config not found.");
                    }
                },
                Parent.AF.CancelToken
            );

        public IPeopleScoDictionaryNew People { get; private set; }

        public void People_CollectionChanged(
            object Sender,
            DictionaryChangedEventArgs<string, string> args
        )
        {
            People.Serialize();
            //var dict = (PeopleScoDictionaryNew)Sender;
            //dict.Serialize();
        }

        //private PeopleScoDictionary _dictPPL;
        //public IPeopleScoDictionary DictPPL => Initialized(_dictPPL, () => LoadDictPPL());
        //private PeopleScoDictionary LoadDictPPL()
        //{
        //    if (Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging))
        //    {
        //        var dictPPL = new PeopleScoDictionary(filename: _defaults.FilenameDictPpl,
        //                                          folderpath: pythonStaging,
        //                                          appGlobals: Parent,
        //                                          prefix: PrefixList.Find(x => x.PrefixType == PrefixTypeEnum.People));

        //        return dictPPL;
        //    }
        //    else { return null; }
        //}
        //async private Task LoadDictPPLAsync() => _dictPPL = await Task.Run(LoadDictPPL);
        //async private Task LoadPrefixAndDictPeopleAsync()
        //{
        //    await LoadPrefixListAsync();
        //    await LoadDictPPLAsync();
        //}

        public string FnameIDList => _defaults.FileName_IDList;

        private IIDList _idList;

        //TODO: Convert IDList to ScoCollection
        public IIDList IDList => Initialized(_idList, () => LoadIDList());

        private async Task LoadIdListAsync()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("AppData", out var appData))
            {
                var outlookApplication = Parent.Ol.App;
                _idList = await Task.Run(() => (IIDList)LoadIdListFromDisk(appData));

                if (_idList.Count == 0 && outlookApplication is not null)
                {
                    _idList.RefreshIDList(outlookApplication);
                }
            }
            else
            {
                _idList = null;
            }
        }

        private IDList LoadIdListFromDisk(string appData)
        {
            var filePath = Path.Combine(appData, FnameIDList);
            List<string> ids;

            try
            {
                ids = FileExists(filePath)
                    ? JsonConvert.DeserializeObject<List<string>>(ReadAllText(filePath)) ?? []
                    : [];
            }
            catch (JsonException)
            {
                ids = [];
            }
            catch (IOException)
            {
                ids = [];
            }

            var idList = new IDList(ids) { Filename = FnameIDList, Folderpath = appData };
            return idList;
        }

        private IIDList LoadIDList()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("AppData", out var appData))
            {
                var idList = new IDList(FnameIDList, appData, Parent.Ol.App);
                if (idList.Count == 0)
                {
                    idList.RefreshIDList();
                }
                return idList;
            }
            else
            {
                return null;
            }
        }

        private string _fnameDictRemap;
        public string FnameDictRemap =>
            Initialized(_fnameDictRemap, () => _fnameDictRemap = _defaults.FileName_DictRemap);

        private ScoDictionaryNew<string, string> _dictRemap;
        public IScoDictionaryNew<string, string> DictRemap =>
            Initialized(_dictRemap, () => LoadDictRemap());

        private ScoDictionaryNew<string, string> LoadDictRemap()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging))
            {
                var dictRemap = ScoDictionaryNew<string, string>.Static.Deserialize(
                    FnameDictRemap,
                    pythonStaging
                );
                return dictRemap;
            }
            else
            {
                return null;
            }
        }

        private async Task LoadDictRemapAsync() =>
            _dictRemap = await Task.Run(LoadDictRemap, default);

        //TODO: Convert CategoryFilters to ScoCollection
        private ISerializableList<string> _catFilters;
        public ISerializableList<string> CategoryFilters
        {
            get =>
                Initialized(
                    _catFilters,
                    () =>
                    {
                        if (
                            Parent.FS.SpecialFolders.TryGetValue(
                                "PythonStaging",
                                out var pythonStaging
                            )
                        )
                        {
                            return new SerializableList<string>(
                                filename: _defaults.FileName_CategoryFilters,
                                folderpath: pythonStaging
                            );
                        }
                        else
                        {
                            return null;
                        }
                    }
                );
            set
            {
                _catFilters = value;

                if (
                    _catFilters.Folderpath == ""
                    && Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging)
                )
                {
                    _catFilters.Folderpath = pythonStaging;
                    _catFilters.Filename = _defaults.FileName_CategoryFilters;
                }
                if (_catFilters.Folderpath != "")
                {
                    _catFilters.Serialize();
                }
            }
        }

        private async Task LoadCategoryFiltersAsync()
        {
            _catFilters = await Task.Run(
                () =>
                {
                    if (
                        Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging)
                    )
                    {
                        return new SerializableList<string>(
                            filename: _defaults.FileName_CategoryFilters,
                            folderpath: pythonStaging
                        );
                    }
                    else
                    {
                        return null;
                    }
                },
                default
            );
        }

        // Prefix List
        private ConcurrentObservableCollection<IPrefix> _prefixList;
        public ConcurrentObservableCollection<IPrefix> PrefixList =>
            Initialized(_prefixList, () => LoadPrefixList());

        public ConcurrentObservableCollection<IPrefix> LoadPrefixList()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging))
            {
                var prefixList = new ConcurrentObservableCollection<IPrefix>(
                    fileName: _defaults.FileName_PrefixList,
                    folderPath: pythonStaging
                );

                if (prefixList.Count == 0)
                {
                    var tdDefaults = new ToDoDefaults();
                    foreach (var prefix in tdDefaults.PrefixList)
                    {
                        prefixList.Add(prefix);
                    }
                    prefixList.Serialize();
                }
                return prefixList;
            }
            else
            {
                return null;
            }
        }

        private async Task LoadPrefixListAsync()
        {
            _prefixList = await Task.Run(LoadPrefixList);
        }

        private ScoDictionaryNew<string, int> _filteredFolderScraping;
        public ScoDictionaryNew<string, int> FilteredFolderScraping =>
            Initialized(_filteredFolderScraping, () => LoadFilteredFolderScraping());

        public ScoDictionaryNew<string, int> LoadFilteredFolderScraping()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging))
            {
                var filteredFolderScraping = ScoDictionaryNew<string, int>.Static.Deserialize(
                    _defaults.FileName_FilteredFolderScraping,
                    pythonStaging
                );
                return filteredFolderScraping;
            }
            else
            {
                return null;
            }
        }

        private async Task LoadFilteredFolderScrapingAsync()
        {
            _filteredFolderScraping = await Task.Run(
                () => LoadFilteredFolderScraping(),
                default(CancellationToken)
            );
        }

        private ScoDictionaryNew<string, string> _folderRemap;
        public ScoDictionaryNew<string, string> FolderRemap =>
            Initializer.GetOrLoad(ref _folderRemap, () => LoadFolderRemap());

        public ScoDictionaryNew<string, string> LoadFolderRemap()
        {
            if (Parent.FS.SpecialFolders.TryGetValue("PythonStaging", out var pythonStaging))
            {
                var folderRemap = ScoDictionaryNew<string, string>.Static.Deserialize(
                    _defaults.FileName_FolderRemap,
                    pythonStaging
                );
                return folderRemap;
            }
            else
            {
                return null;
            }
        }

        private async Task LoadFolderRemapAsync()
        {
            _folderRemap = await Task.Run(LoadFolderRemap);
        }

        public Func<IEnumerable<string>, IPrefix, string, string, string> FindMatchingTag
        {
            get;
            internal set;
        } = TagLauncher.LaunchAndFindMatch;

        //IEnumerable<string> options, IApplicationGlobals appGlobals
        public Func<IEnumerable<string>, List<string>> SelectFromList { get; internal set; }

        private async Task LoadSelectFromListAsync()
        {
            await Task.Run(() =>
                SelectFromList = (options) => TagLauncher.LaunchAndSelect(options, this.Parent)
            );
        }

        public IFlagChangeTrainingQueue FlagChangeTrainingQueue { get; set; }

        private async Task LoadFlagChangeTrainingQueueAsync()
        {
            await Task.Run(() =>
            {
                FlagChangeTrainingQueue = new FlagChangeTrainingQueue().Init();
            });
        }
    }
}

File: TaskMasterOld/AppGlobals/AppToDoObjects.vb
Imports Newtonsoft.Json
Imports System.IO
Imports ToDoModel
Imports UtilitiesVB
Imports UtilitiesCS

Public Class AppToDoObjects
    Implements IToDoObjects

    Private _projInfo As ProjectInfo
    Private _dictPPL As Dictionary(Of String, String)
    Private _IDList As ListOfIDs
    Private ReadOnly _parent As ApplicationGlobals
    Private _dictRemap As Dictionary(Of String, String)
    Private _catFilters As SerializableList(Of String)

    Public Sub New(ParentInstance As ApplicationGlobals)
        _parent = ParentInstance
    End Sub

    Public ReadOnly Property Parent As IApplicationGlobals Implements IToDoObjects.Parent
        Get
            Return _parent
        End Get
    End Property

    Public ReadOnly Property ProjInfo_Filename As String Implements IToDoObjects.ProjInfo_Filename
        Get
            Return My.Settings.FileName_ProjInfo
        End Get
    End Property

    Public ReadOnly Property ProjInfo As IProjectInfo Implements IToDoObjects.ProjInfo
        Get
            If _projInfo Is Nothing Then
                _projInfo = LoadToDoProjectInfo(Path.Combine(Parent.FS.FldrAppData, My.Settings.FileName_ProjInfo))
            End If
            Return _projInfo
        End Get
    End Property


    Public ReadOnly Property DictPPL_Filename As String Implements IToDoObjects.DictPPL_Filename
        Get
            Return My.Settings.FilenameDictPpl
        End Get
    End Property

    Public ReadOnly Property DictPPL As Dictionary(Of String, String) Implements IToDoObjects.DictPPL
        Get
            If _dictPPL Is Nothing Then
                _dictPPL = LoadDictJSON(Parent.FS.FldrStaging, DictPPL_Filename)
            End If
            Return _dictPPL
        End Get
    End Property

    Public Sub DictPPL_Save() Implements IToDoObjects.DictPPL_Save
        File.WriteAllText(
            Path.Combine(Parent.FS.FldrStaging, DictPPL_Filename),
            JsonConvert.SerializeObject(_dictPPL, Formatting.Indented))
    End Sub

    Public ReadOnly Property FnameIDList As String Implements IToDoObjects.FnameIDList
        Get
            Return My.Settings.FileName_IDList
        End Get
    End Property

    Public ReadOnly Property IDList As IListOfIDs Implements IToDoObjects.IDList
        Get
            If _IDList Is Nothing Then
                _IDList = New ListOfIDs(
                    Path.Combine(Parent.FS.FldrAppData,
                                 My.Settings.FileName_IDList),
                    _parent.Ol.App)
            End If
            Return _IDList
        End Get
    End Property

    Public ReadOnly Property FnameDictRemap As String Implements IToDoObjects.FnameDictRemap
        Get
            Return My.Settings.FileName_DictRemap
        End Get
    End Property

    Public ReadOnly Property DictRemap As Dictionary(Of String, String) Implements IToDoObjects.DictRemap
        Get
            If _dictRemap Is Nothing Then
                _dictRemap = LoadDictCSV(Parent.FS.FldrStaging, My.Settings.FileName_DictRemap)
            End If
            Return _dictRemap
        End Get
    End Property

    Public ReadOnly Property CategoryFilters As ISerializableList(Of String) Implements IToDoObjects.CategoryFilters
        Get
            If _catFilters Is Nothing Then
                Dim _catFilters = New SerializableList(Of String)
                With _catFilters
                    .Filename = My.Settings.FileName_CategoryFilters
                    .Folderpath = Parent.FS.FldrAppData
                    If File.Exists(.Folderpath) Then
                        .Deserialize()
                    Else
                        Dim tempList = New SerializableList(Of String)(CCOCatList_Load())
                        tempList.Folderpath = .Folderpath
                        _catFilters = tempList
                        .Serialize()
                    End If
                End With
            End If
            Return _catFilters
        End Get
    End Property

    Private Function LoadDictCSV(fpath As String,
                                 filename As String) _
                                 As Dictionary(Of String, String)
        Dim dict As Dictionary(Of String, String) = UtilitiesVB.LoadDictCSV(fpath, filename.Split(".")(0) & ".csv")
        If dict IsNot Nothing Then WriteDictJSON(dict, Path.Combine(fpath, filename))
        Return dict
    End Function

    Private Function LoadDictJSON(fpath As String,
                                  filename As String) _
                                  As Dictionary(Of String, String)

        Dim filepath As String = Path.Combine(fpath, filename)
        Dim dict As Dictionary(Of String, String) = Nothing
        Dim response As MsgBoxResult = MsgBoxResult.Ignore

        Try
            dict = JsonConvert.DeserializeObject(Of Dictionary(Of String, String)) _
                    (File.ReadAllText(Path.Combine(Parent.FS.FldrStaging, DictPPL_Filename)))
        Catch ex As FileNotFoundException
            response = MsgBox(filepath & "not found. Load from CSV?", vbYesNo)
        Catch ex As System.Exception
            response = MsgBox(filepath & "encountered a problem. " & ex.Message & "Load from CSV?", vbYesNo)
        Finally
            If response = vbYes Then
                dict = LoadDictCSV(fpath, filename)
            ElseIf response = vbNo Then
                response = MsgBox("Start a new blank dictionary?", vbYesNo)
                If response = vbYes Then
                    dict = New Dictionary(Of String, String)
                Else
                    Throw New ArgumentNullException("Cannot proceed without dictionary: " & filename)
                End If
            End If
        End Try
        Return dict
    End Function

    Public Sub WriteDictJSON(dict As Dictionary(Of String, String), filepath As String)
        File.WriteAllText(filepath, JsonConvert.SerializeObject(dict, Formatting.Indented))
    End Sub
End Class
